libs/libzauth/libzauth-c/src/lib.rs

```diff
@@ -284,8 +312,191 @@ impl From<TokenVerification> for ZauthTokenVerification {
     fn from(t: TokenVerification) -> ZauthTokenVerification {
         match t {
             TokenVerification::Verified => ZauthTokenVerification::Verified,
-            TokenVerification::Invalid  => ZauthTokenVerification::Invalid,
-            TokenVerification::Pending  => ZauthTokenVerification::Pending,
+            TokenVerification::Invalid => ZauthTokenVerification::Invalid,
+            TokenVerification::Pending => ZauthTokenVerification::Pending,
+        }
+    }
+}
+
+// ------------------------------------------------------------------------
+// OAuth
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug)]
+pub enum OAuthResultStatus {
+    Ok = 0,
+    InsufficientScope = 1,
+    NullArg = 2,
+    IoError = 3,
+    Utf8Error = 4,
+    Panic = 99,
+}
+
+pub struct OAuthJwk(String);
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug)]
+pub struct OAuthResult {
+    uid: *const libc::c_char,
+    status: OAuthResultStatus,
+}
+
+fn catch_unwind_with<F, R>(f: F, r: R) -> R
+where
+    F: FnOnce() -> R + UnwindSafe,
+{
+    match panic::catch_unwind(f) {
+        Ok(x) => x,
+        Err(_) => r,
+    }
+}
+
+impl From<OauthError> for OAuthResultStatus {
+    fn from(e: OauthError) -> OAuthResultStatus {
+        match e {
+            OauthError::JsonError(_) => Self::Panic,
+            OauthError::JwtSimpleError(_) => Self::Panic,
+            OauthError::Base64DecodeError(_) => Self::Panic,
+            OauthError::InvalidJwk => Self::Panic,
+            OauthError::InvalidJwtNoSubject => Self::Panic,
+            OauthError::InvalidScope => Self::InsufficientScope,
+        }
+    }
+}
+
+impl From<io::Error> for OAuthResultStatus {
```

Comment:
You probably could have handled all these manual `From` implementations with the `thiserror` crate.
But it could be tricky to handle with the discriminant so leave it alone if it's too complex

```rust
#[repr(C)]
#[derive(Error, Clone, Copy, Debug)]
pub enum OAuthResultStatus {
    Ok,
    IoError(#[from] io::Error),
}
```